homeassistant/components/apple_tv/__init__.py

```diff
@@ -70,205 +70,257 @@ def ensure_list(value: Union[T, Sequence[T]]) -> Sequence[T]:
-        entity_ids = service.data.get(ATTR_ENTITY_ID)
-
-        if service.service == SERVICE_SCAN:
-            hass.async_add_job(scan_apple_tvs, hass)
-            return
-
-        if entity_ids:
-            devices = [
-                device
-                for device in hass.data[DATA_ENTITIES]
-                if device.entity_id in entity_ids
-            ]
-        else:
-            devices = hass.data[DATA_ENTITIES]
-
-        for device in devices:
-            if service.service != SERVICE_AUTHENTICATE:
-                continue
-
-            atv = device.atv
-            credentials = await atv.airplay.generate_credentials()
-            await atv.airplay.load_credentials(credentials)
-            _LOGGER.debug("Generated new credentials: %s", credentials)
-            await atv.airplay.start_authentication()
-            hass.async_add_job(request_configuration, hass, config, atv, credentials)
-
-    async def atv_discovered(service, info):
-        """Set up an Apple TV that was auto discovered."""
-        await _setup_atv(
-            hass,
-            config,
-            {
-                CONF_NAME: info["name"],
-                CONF_HOST: info["host"],
-                CONF_LOGIN_ID: info["properties"]["hG"],
-                CONF_START_OFF: False,
-            },
-        )
+    An instance is used per device to share the same power state between
+    several platforms. It also manages scanning and connection establishment
+    in case of problems.
+    """
 
-    discovery.async_listen(hass, SERVICE_APPLE_TV, atv_discovered)
+    def __init__(self, hass, config_entry):
+        """Initialize power manager."""
+        self.config_entry = config_entry
+        self.hass = hass
+        self.listeners = []
+        self.message = None
+        self.atv = None
+        self._is_on = not config_entry.options.get(CONF_START_OFF, False)
+        self._connection_attempts = 0
+        self._connection_was_lost = False
+        self._task = None
+
+    async def init(self):
+        """Initialize power management."""
+        if self._is_on:
+            await self.connect()
+
+    def connection_lost(self, exception):
+        """Device was unexpectedly disconnected."""
+        _LOGGER.warning('Connection lost to Apple TV "%s"', self.atv.name)
+
+        self.atv = None
+        self._connection_was_lost = True
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
+
+    def connection_closed(self):
+        """Device connection was (intentionally) closed."""
+        self.atv = None
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
+
+    async def connect(self):
+        """Connect to device."""
+        self._is_on = True
+        self._start_connect_loop()
+
+    async def disconnect(self):
+        """Disconnect from device."""
+        _LOGGER.debug("Disconnecting from device")
+        self._is_on = False
+        try:
```

Comment:
Try without except?

Reply:
Yeah, there's not much I can do in case this fails but I want to make sure that the media player state is updated with some useful information to the user and keep internal state sane (thus the `finally` block at the end). I could catch "something" and log that, but it's better to fail hard.

Reply:
I have to disagree, IMHO it is always better to at least catch something an log it to the user. Providing the user with more information and context of the occurred error. Furthermore, a hard fail should always be prevented.

Reply:
Yes, in principal I agree with you but in this case it will always be logged to the user with complete stack trace by the default exception handler. So no information is lost. In addition, this particular case should *never* raise an exception. If that happens, it's bad and in general it's better to fail fast and hard in order to find the problem (e.g. via problem report) and fix it. But I guess we need a second opinion.

Reply:
Yeah I agree, a second opinion on this topic would be great.

Reply:
I added catching of exceptions here, so we should be good now.